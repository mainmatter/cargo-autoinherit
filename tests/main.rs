use std::{env::temp_dir, fs, path::PathBuf};

use cargo_autoinherit::{auto_inherit, AutoInheritConf};
use git2::Repository;
use insta::assert_snapshot;
use rand::{
    distributions::{Alphanumeric, DistString},
    thread_rng,
};

const TEST_REPO_URL: &str = match option_env!("TEST_REPO_URL") {
    Some(url) => url,
    None => "https://github.com/hdoordt/cargo-autoinherit-test",
};

fn clone_test_project() -> PathBuf {
    let dir: String = Alphanumeric.sample_string(&mut thread_rng(), 32);
    let dir = temp_dir().join(dir);
    Repository::clone(TEST_REPO_URL, &dir).expect("Error cloning test repository");
    dir
}

fn run_with_conf(conf: AutoInheritConf) -> PathBuf {
    let repo_path = clone_test_project();

    let conf = AutoInheritConf {
        manifest_path: Some(PathBuf::from(format!(
            "{}/Cargo.toml",
            repo_path.to_str().unwrap()
        ))),
        ..conf
    };
    auto_inherit(conf).unwrap();
    repo_path
}

#[test]
fn default_behavior() {
    let repo_path = run_with_conf(AutoInheritConf::default());
    insta::glob!(repo_path, "**/Cargo.toml", |p| {
        let input = fs::read_to_string(dbg!(p)).unwrap();
        assert_snapshot!(input);
    });
}

#[test]
fn prefer_simple_dotted() {
    let repo_path = run_with_conf(AutoInheritConf {
        prefer_simple_dotted: true,
        ..Default::default()
    });
    insta::glob!(repo_path, "**/Cargo.toml", |p| {
        let input = fs::read_to_string(dbg!(p)).unwrap();
        assert_snapshot!(input);
    });
}

#[test]
fn with_excluded_member() {
    let repo_path = run_with_conf(AutoInheritConf {
        prefer_simple_dotted: false,
        exclude_members: vec!["cargo-autoinherit-test-cli".to_string()],
        ..Default::default()
    });
    insta::glob!(repo_path, "**/Cargo.toml", |p| {
        let input = fs::read_to_string(dbg!(p)).unwrap();
        assert_snapshot!(input);
    });
}
