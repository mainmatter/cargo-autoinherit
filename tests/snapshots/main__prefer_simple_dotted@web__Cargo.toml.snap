---
source: tests/main.rs
expression: input
snapshot_kind: text
---
[package]
name = "cargo-autoinherit-test-web"
version = "0.0.1"
edition = "2021"
publish = false

[lib]
# examples in docs don't run without a running database, etc.
doctest = false

[features]
test-helpers = ["dep:serde_json", "dep:tower", "dep:hyper"]

[dependencies]
anyhow = "1.0"
axum = { version = "0.7", features = ["macros"] }
cargo-autoinherit-test-config = { path = "../config" }
cargo-autoinherit-test-db = { path = "../db", features = ["openapi"] }
serde = { version = "1.0", features = ["derive"] }
tokio = { version = "1.34", features = ["full"] }
tower-http = { version = "0.6", features = ["full"] }
tracing = "0.1"
tracing-panic = "0.1"
tracing-subscriber = { version = "0.3", features = ["env-filter", "registry", "fmt"] }
uuid = { version = "1.6", features = ["serde"] }
serde_json = { version = "1.0", optional = true }
thiserror = "1.0"
tower = { version = "0.5", features = ["util"], optional = true }
hyper = { version = "1.0", features = ["full"], optional = true }
cargo-autoinherit-test-macros = { path = "../macros" }
validator = { version = "0.18.1", features = ["derive"] }
utoipa = { version = "5.1.2", features = ["axum_extras", "uuid"] }
utoipa-axum = {version = "0.1.2" }
utoipa-swagger-ui = { version = "8.0.3", features = ["axum", "reqwest"] }

[dev-dependencies]
fake = "2.9"
googletest = "0.12"
cargo-autoinherit-test-db = { path = "../db", features = ["test-helpers"] }
cargo-autoinherit-test-web = { path = ".", features = ["test-helpers"] }
